text/0000-proc-macros.md

```diff
@@ -0,0 +1,452 @@
+
+The details of libmacro will be specified in a future RFC. In the meantime, this
+[blog post](http://ncameron.org/blog/libmacro/) gives an idea of what it might
+contain.
+
+The philosophy here is that libmacro will contain low-level tools for
+constructing macros, dealing with tokens, hygiene, pattern matching, quasi-
+quoting, interactions with the compiler, etc. For higher level abstractions
+(such as parsing and an AST), macros should use external libraries (there are no
+restrictions on `#[cfg(macro)]` crates using other crates).
+
+The `MacroContext` is an object passed to all procedural macro definitions. It
+is the main entry point to the libmacro API and for interaction with the
+compiler. Via the `MacroContext`, a procedural macro can access information
+about the context in which it is used and defined, and perform operations which
+rely on the state of the compiler. It will be more fully defined in the upcoming
+RFC proposing libmacro.
+
+Rust macros are hygienic by default. Hygiene is a large and complex subject, but
+to summarise: effectively, naming takes place in the context of the macro
+definition, not the expanded macro.
+
+Procedural macros often want to bend the rules around macro hygiene, for example
+to make items or variables more widely nameable than they would be by default.
+Procedural macros will be able to take part in the application of the hygiene
+algorithm via libmacro. Again, full details must wait for the libmacro RFC and a
+sketch is available in this [blog post](http://ncameron.org/blog/libmacro/).
+
+
+## Tokens
+
+Procedural macros will primarily operate on tokens. There are two main benefits
+to this principle: flexibility and future proofing. By operating on tokens, code
+passed to procedural macros does not need to satisfy the Rust parser, only the
+lexer. Stabilising an interface based on tokens means we need only commit to
+not changing the rules around those tokens, not the whole grammar. I.e., it
+allows us to change the Rust grammar without breaking procedural macros.
+
+In order to make the token-based interface even more flexible and future-proof,
+I propose a simpler token abstraction than is currently used in the compiler.
+The proposed system may be used directly in the compiler or may be an interface
+wrapper over a more efficient representation.
+
+Since macro expansion will not operate purely on tokens, we must keep hygiene
+information on tokens, rather than on `Ident` AST nodes (we might be able to
+optimise by not keeping such info for all tokens, but that is an implementation
+detail). We will also keep span information for each token, since that is where
+a record of macro expansion is maintained (and it will make life easier for
+tools. Again, we might optimise internally).
+
+A token is a single lexical element, for example, a numeric literal, a word
+(which could be an identifier or keyword), a string literal, or a comment.
```

Comment:
In the current compiler, you can do that by looking up the source between adjacent token's spans.